rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,385 @@
+            MyWidgetContext {
+                widget: self,
+                context,
+            }
+        }
+    }
+
+    impl Widget for MyWidget {
+        fn get_id(&self) -> Index {
+            self.id
+        }
+
+        fn focusable(&self) -> bool {
+            true
+        }
+
+        fn set_id(&mut self, id: Index) {
+            self.id = id;
+        }
+
+        fn get_styles(&self) -> Option<Style> {
+            self.styles.clone()
+        }
+
+        fn get_name(&self) -> String {
+            String::from("MyWidget")
+        }
+
+        fn on_event(&mut self, context: &mut KayakContext, event: &mut crate::Event) {
+            let this = self.generate_context(context);
+            if let Some(on_event) = this.widget.on_event.as_mut() {
+                if let Ok(mut on_event) = on_event.0.write() {
+                    on_event(this.context, event);
+                }
+            }
+        }
+
+        fn render(&mut self, context: &mut KayakContext) {
+            Self::render(self.generate_context(context));
+        }
+    }
+}
+
+use my_widget_module::{MyWidget, MyWidgetContext, MyWidgetProps};
+// Outside the module so it doesn't have access to the internals
+impl MyWidget {
+    pub fn render(mut this: MyWidgetContext) {
+        let MyWidgetProps { fill, name } = this.props();
+
+        this.set_styles(Some(Style {
+            render_command: StyleProp::Value(RenderCommand::Empty),
+            ..Default::default()
+        }));
+
+        this.on_event(Some(OnEvent::new(move |_, evt| {
+            println!("Event fired!");
+        })));
+
+        // ...
+    }
+}
+```
+
+</details>
+
+## Feedback
+
+This is all just me thinking of ways to improve on these issues. Are they worth the trouble? And would we stand to gain from implementing a solution like this? I think so, but I could be totally and completely wrong. There could also be other, better solutions than the one I came up with. 
+
+Anyways, just wanted to bring this up and see what people thought. So let me know!
+
+## Addendum
+
+### Tracking Changes
+
+An additional feature this might allow is tracking changes that can't be tracked during the render phase:
+
+```rust
+#[widget]
+fn MyWidget(disabled: bool) {
+  this.set_focusable(disabled);
+  // ...
+}
+```
+And in `MyWidgetContext`:
+
+```rust
+pub fn set_focusable(&mut self, focusable: bool) {
+  self.widget.focusable = focusable;
+
+  // If we need to track this change, we could do something like:
+  self.edits.insert(WidgetEdit::Focus);
+}
+```
+
+Then if we need to handle any changes we can just iterate through the set of changes and handle them individually.
+
+```rust
+impl Widget for MyWidget {
+  // ...
+  fn on_event(&mut self, context: &mut KayakContext, event: &mut crate::Event) {
+    let mut this = self.generate_context(context);
+    if let Some(mut on_event) = this.widget.on_event.clone() {
+      if let Ok(mut on_event) = on_event.0.write() {
+        on_event(this, event);
```

Comment:
Actually this should be:

```suggestion
        on_event(&mut this, event);
```

Since we want to read from it after passing the event.